### ISSUE TITLE
Better support for overlapping parameters

### ISSUE THREAD

**Issue by ibraheemdev** [ibraheemdev/matchit#13]
Currently routes like `/:a/:b` and `/:a` are supported, but only if they have a common prefix (`/:a`). This means a route like `/:a/:b` conflicts with `/:c`. We can support this by normalizing route parameters positionally when they are first inserted. `/:user/:id` would become `/:a/:b`, and `/:id` becomes `/:a`. Param nodes get an additional field representing the user-defined parameter name, which is used to construct the parameter list.

**Comment by BratSinot** [comment]
Also, another variant:
```rust
use axum::{
    routing::{get, post},
    Router,
};

fn main() {
    /* these are configurable */
    let base_path = "/";
    let get_base_path = "/";
    let post_base_path = "/";

    let get_route = Router::new().nest(
        get_base_path,
        Router::new().route("/get", get(|| std::future::ready("Ho".to_owned()))),
    );
    let post_route = Router::new().nest(
        post_base_path,
        Router::new().route("/post", post(|| std::future::ready("Ho".to_owned()))),
    );
    let app: Router<String> =
        Router::new().nest(base_path, Router::new().merge(get_route).merge(post_route));
}
```
cause this error:
```log
thread 'main' panicked at 'Invalid route "/": insertion failed due to conflict with previously registered route: /*__private__axum_nest_tail_param', src/main.rs:21:70
stack backtrace:
```

**Comment by ibraheemdev** [comment]
@BratSinot I'm not 100% familiar with how axum uses matchit internally, but that looks more like an axum issue than a matchit one. Maybe open an axum issue?

**Comment by BratSinot** [comment]
> @BratSinot I'm not 100% familiar with how axum uses matchit internally, but that looks more like an axum issue than a matchit one. Maybe open an axum issue?

Ow, sorry! I came here from `axum` issues list =)

**Comment by ibraheemdev** [comment]
I realized this is more complicated because the name of the parameter would change depending on the rest of the path, which hasn't been parsed yet. There's another approach that could work by storing a remapping at the end of each route, but I'm starting to think it's not worth handling this edge case.

**Comment by ibraheemdev** [comment]
Storing the remapping at leaf nodes turned out to be simpler than expected. I've implemented it in https://github.com/ibraheemdev/matchit/pull/37, this should be released as part of 0.8.0.

**Comment by davidpdrsn** [comment]
Any change you would back port that to 0.7 as well? I think axum users would really appreciate that :)

**Comment by ibraheemdev** [comment]
Released in 0.7.3 :tada:

**Comment by davidpdrsn** [comment]
Great news! Thanks for fixing it 😃
